src/box/lua/net_box.c

```diff
@@ -3187,14 +3187,11 @@ luaT_netbox_transport_stop(struct lua_State *L)
 		box_error_raise(ER_NO_CONNECTION, "Connection closed");
 		netbox_transport_set_error(transport);
 		transport->state = NETBOX_CLOSED;
-		netbox_transport_on_state_change(transport, L);
+		netbox_transport_on_state_change_pcall(transport, L);
 	}
 	/* Cancel the worker fiber. */
-	if (transport->worker != NULL) {
+	if (transport->worker != NULL)
 		fiber_cancel(transport->worker);
-		/* Check if we cancelled ourselves. */
-		luaL_testcancel(L);
```

Comment:
TBH I don't understand this change. Why was `luaL_testcancel()` needed before and why are we sure removing it is safe now?

Reply:
We need to make this change, because we are currently throwing an exception from code that is called as a cdata finalizer, which is considered undefined behaviour, and this code is broken anyways.

AFAIU, this code aimed to solve the scenario when `connection:close` is called from a connection trigger (which is the only way how a connection can be closed from the worker fiber other than being garbage collected while entering the Lua VM from the worker fiber). However, it fails to solve even this scenario, since one can `pcall` the `connection:close` call and the exception won't stop the worker fiber.

Reply:
> AFAIU, this code aimed to solve the scenario when connection:close is called from a connection trigger (which is the only way how a connection can be closed from the worker fiber other than being garbage collected while entering the Lua VM from the worker fiber). However, it fails to solve even this scenario, since one can pcall the connection:close call and the exception won't stop the worker fiber.

I see, thanks.

Reply:
`stop` is called under `pcall` from `gc` so I don't see how it can result in UB.

Reply:
My bad, I overlooked this code, thought we are dealing with this code:
https://github.com/tarantool/tarantool/blob/58f5ffe43bffa0901d8d577bbb4d523c4e7e682a/src/box/lua/net_box.lua#L419-L439

Looks like there is no UB after all, so I am dropping this patch and closing the ticket.

Reply:
@locker, thanks for the catch!